tpm2/direct/structures.go

```diff
@@ -0,0 +1,1237 @@
+// See definition in Part 2, Structures, section 5.3.
+type TPMModifierIndicator uint32
+
+// TPMAuthorizationSize represents a TPM_AUTHORIZATION_SIZE.
+// the authorizationSize parameter in a command
+// See definition in Part 2, Structures, section 5.3.
+type TPMAuthorizationSize uint32
+
+// TPMParameterSize represents a TPM_PARAMETER_SIZE.
+// the parameterSize parameter in a command
+// See definition in Part 2, Structures, section 5.3.
+type TPMParameterSize uint32
+
+// TPMKeySize represents a TPM_KEY_SIZE.
+// a key size in octets
+// See definition in Part 2, Structures, section 5.3.
+type TPMKeySize uint16
+
+// TPMKeyBits represents a TPM_KEY_BITS.
+// a key size in bits
+// See definition in Part 2, Structures, section 5.3.
+type TPMKeyBits uint16
+
+// TPMGenerated represents a TPM_GENERATED.
+// See definition in Part 2: Structures, section 6.2.
+type TPMGenerated uint32
+
+// Check verifies that a TPMGenerated value is correct, and returns an error
+// otherwise.
+func (g TPMGenerated) Check() error {
+	if g != TPMGeneratedValue {
+		return fmt.Errorf("TPM_GENERATED value should be 0x%x, was 0x%x", TPMGeneratedValue, g)
+	}
+	return nil
+}
+
+// TPMAlgID represents a TPM_ALG_ID.
+// See definition in Part 2: Structures, section 6.3.
+type TPMAlgID uint16
+
+// TPMECCCurve represents a TPM_ECC_Curve.
+// See definition in Part 2: Structures, section 6.4.
+type TPMECCCurve uint16
+
+// TPMCC represents a TPM_CC.
+// See definition in Part 2: Structures, section 6.5.2.
+type TPMCC uint32
+
+// TPMRC represents a TPM_RC.
+// See definition in Part 2: Structures, section 6.6.
+type TPMRC uint32
+
+// TPMST represents a TPM_ST.
+// See definition in Part 2: Structures, section 6.9.
+type TPMST uint16
+
+// TPMSE represents a TPM_SE.
+// See definition in Part 2: Structures, section 6.11.
+type TPMSE uint8
+
+// TPMCap represents a TPM_CAP.
+// See definition in Part 2: Structures, section 6.12.
+type TPMCap uint32
+
+// TPMPT represents a TPM_PT.
+// See definition in Part 2: Structures, section 6.13.
+type TPMPT uint32
+
+// TPMPTPCR represents a TPM_PT_PCR.
+// See definition in Part 2: Structures, section 6.14.
+type TPMPTPCR uint32
+
+// TPMHandle represents a TPM_HANDLE.
+// See definition in Part 2: Structures, section 7.1.
+type TPMHandle uint32
+
+// TPMAAlgorithm represents a TPMA_ALGORITHM.
+// See definition in Part 2: Structures, section 8.2.
+type TPMAAlgorithm struct {
+	// SET (1): an asymmetric algorithm with public and private portions
+	// CLEAR (0): not an asymmetric algorithm
+	Asymmetric bool `gotpm:"bit=0"`
+	// SET (1): a symmetric block cipher
+	// CLEAR (0): not a symmetric block cipher
+	Symmetric bool `gotpm:"bit=1"`
+	// SET (1): a hash algorithm
+	// CLEAR (0): not a hash algorithm
+	Hash bool `gotpm:"bit=2"`
+	// SET (1): an algorithm that may be used as an object type
+	// CLEAR (0): an algorithm that is not used as an object type
+	Object    bool  `gotpm:"bit=3"`
+	Reserved1 uint8 `gotpm:"bit=7:4"`
```

Comment:
Open question: should we omit this from the definition, and check that reserved bits are all zeros?

Reply:
Resolved with @josephlr's suggestion to embed an unexported type whose job it is to store reserved bits.